src/Database/Dpi/Internal.chs

```diff
@@ -624,111 +624,106 @@ newtype Data = Data (NativeTypeNum -> OracleTypeNum -> IO DataValue)
 instance Storable Data where
   sizeOf    _ = {#sizeof  Data #}
   alignment _ = {#alignof Data #}
-  poke      p' (Data f) = do
-    f NativeTypeDouble OracleTypeBoolean >>= go p'
+  poke p (Data f) = do
+    dataValue <- f NativeTypeDouble OracleTypeBoolean
+    case dataValue of
+      (DataNull          _) -> {#set Data -> isNull #} p 1
+      (DataBoolean       v) -> libDataSetBool p (fromBool v)
+      (DataIntervalDs Data_IntervalDS{..}) -> do
+        libDataSetIntervalDS
+          p
+          (fromIntegral days)
+          (fromIntegral hours)
+          (fromIntegral minutes)
+          (fromIntegral seconds)
+          (fromIntegral fseconds)
+      (DataIntervalYm  Data_IntervalYM {..}) -> libDataSetIntervalYM p years months
+      (DataChar          v) -> sByt p v
+      (DataLongRaw       v) -> sByt p v
+      (DataLongVarchar   v) -> sByt p v
+      (DataNChar         v) -> sByt p v
+      (DataNumBytes      v) -> sByt p v
+      (DataNVarchar      v) -> sByt p v
+      (DataVarchar       v) -> sByt p v
+      (DataRaw           v) -> sByt p v
+      (DataTimestamp     v) -> sTmp p v
+      (DataTimestampLtz  v) -> sTmp p v
+      (DataTimestampTz   v) -> sTmp p v
+      (DataDate          v) -> sTmp p v
+      (DataBFile         v) -> libDataSetLOB p v
+      (DataNClob         v) -> libDataSetLOB p v
+      (DataBlob          v) -> libDataSetLOB p v
+      (DataClob          v) -> libDataSetLOB p v
+      (DataDouble        v) -> libDataSetDouble p v
+      (DataFloat         v) -> libDataSetFloat  p v
+      (DataInt           v) -> libDataSetInt64  p (fromIntegral v)
+      (DataUint          v) -> libDataSetUint64 p (fromIntegral v)
+      (DataNumDouble     v) -> libDataSetDouble p (realToFrac v)
+      (DataNumInt        v) -> libDataSetInt64  p (fromIntegral v)
+      (DataNumUint       v) -> libDataSetUint64 p (fromIntegral v)
+      (DataObject        v) -> libDataSetObject p v
+      (DataRowid         v) -> do
+        {#set Data -> isNull #} p 0
+        {#set Data -> value.asRowid  #} p v
+      (DataStmt          v) -> libDataSetStmt   p v
+      (DataTimestampD    v) -> libDataSetDouble p v
+      (DataTimestampLtzD v) -> libDataSetDouble p v
+      (DataTimestampTzD  v) -> libDataSetDouble p v
     where
-      sLob p1 v1 = {#set Data -> value.asLOB        #} p1 v1
-      sByt p2 Data_Bytes{..} = do
+      sByt p' Data_Bytes{..} = do
+        {#set Data -> isNull #} p' 0
```

Comment:
where we don't use the lib* functions, we need to explicitly be sure to set isNull to 0